src/alire/alire-origins-deployers-system-homebrew.adb

```diff
@@ -0,0 +1,273 @@
+            end if;
+            Version := Result;
+         end Get_Installed_Version;
+
+         procedure Info_Callback (Name : UTF8_String; Value : JSON_Value) is
+            function "+"
+              (L : Ada.Strings.Unbounded.Unbounded_String) return String
+              renames Ada.Strings.Unbounded.To_String;
+         begin
+            if Name = "versions" then
+               if Kind (Value) /= JSON_Object_Type then
+                  raise JSON_Issue with "JSON 'versions' not JSON_Object";
+               end if;
+               Get_Available_Version (Value, Available_Version);
+               Trace.Debug ("available: " & (+Available_Version));
+            elsif Name = "installed" then
+               if Kind (Value) /= JSON_Array_Type then
+                  raise JSON_Issue with "JSON 'installed' not JSON_Array";
+               end if;
+               Get_Installed_Version (JSON_Array'(Get (Value)),
+                                      Installed_Version);
+               Trace.Debug ("installed: " & (+Installed_Version));
+            end if;
+         end Info_Callback;
+
+         Arr : JSON_Array;
+      begin
+         if Kind (Data) /= JSON_Array_Type then
+            raise JSON_Issue with "JSON info not JSON_Array";
+         end if;
+         Arr := Get (Data);
+         if Length (Arr) /= 1 then
+            raise JSON_Issue with "JSON info length /= 1";
+         end if;
+         if Kind (Get (Arr, 1)) /= JSON_Object_Type then
+            raise JSON_Issue with "JSON info (1) not JSON_Object";
+         end if;
+         Map_JSON_Object (Get (Arr, 1), Info_Callback'Access);
+      end;
+
+   --  exception
+   --     when E : JSON_Issue =>
+   --     Trace.Debug ("JSON issue: " & Ada.Exceptions.Exception_Message (E));
+   --        Available_Version := Null_Unbounded_String;
+   --        Installed_Version := Null_Unbounded_String;
+   end Get_Info;
+
+   -----------------------
+   -- Already_Installed --
+   -----------------------
+
+   overriding function Already_Installed (This : Deployer) return Boolean
+   is
+   begin
+      Trace.Debug ("already_installed? " & This.Base.Package_Name);
+      if Homebrew_Present then
+         declare
+            Cask_Name : constant String
+              := Homebrew_Prefix & "/Cellar/" & This.Base.Package_Name;
+            use Ada.Directories;
+         begin
+            return
+              Exists (Cask_Name)
+              and then Kind (Cask_Name) = Directory;
+         end;
+      else
+         Trace.Debug ("homebrew not found");
+         return False;
+      end if;
+   end Already_Installed;
+
+   ------------
+   -- Detect --
+   ------------
+
+   overriding
+   function Detect (This : Deployer) return Version_Outcomes.Outcome
+   is
+
+      function Get_Version_From_String (Candidate : String)
+                                       return Version_Outcomes.Outcome;
+      function Get_Version_From_String (Candidate : String)
```

Comment:
The `Semantic_Versioning` library is already providing this kind of conversion. Is there a reason not to use it?

Reply:
No. The existing code will (try to) detect format errors, is it OK to let `Semantic_Versioning` raise an exception in this case? would said exception need to be handled? (it doesn't seem to be handled in other uses).

Reply:
@mosteo I think we need your input here.

Reply:
The extra build part should not be a problem. We have to be careful if distributions use `-` in a sense different than semantic versioning pre-releases, though, to avoid mis-reporting versions.

These detectors shouldn't raise; in the worst case we prefer the external to be reported as undetected. But then, relaxed semantic versioning will accept anything that starts with a number as a version.